### PR TITLE
Test with Symfony 4 components

### PR DESCRIPTION
**Pull request by dbu**
continuation of #389

### CODE REVIEW

**Comment by dbu** [comment]
ping @JonEastman @pgnd if one of you has time to look into those issues, please do (the build in question is the dev build that uses symfony 4. note that its currently allowed to fail, which is the reason why this PR is marked as green)

**Comment by pgnd** [comment]
on a new, clean install of SF 4.0/*release*

with
```
egrep "guzzle|dispatcher" composer.json
	"guzzlehttp/psr7": "^1.4",
	"php-http/guzzle6-adapter": "^1.1",
	"symfony/event-dispatcher": "^4.0",
```
local branch
```
git branch -a | grep \*
	* symfony-4-updated

git log | head
	commit b53772792ee0b88f7e28b253b9813566b4797f93
	Author: David Buchmann <david.buchmann@liip.ch>
	Date:   Thu Nov 30 12:54:13 2017 +0100

	    stop using process builder as its removed in symfony 4

	commit 07c6bb4dd92ee299ac4ab5e871802c0873bec533
	Author: David Buchmann <david.buchmann@liip.ch>
	Date:   Thu Nov 30 11:50:50 2017 +0100
```
(NOTE: this^ is *without* the manual application of the JonEastman/symfony-4-updated PR ...)

is, atm, uninstallable
<details>
<summary>composer -v require friendsofsymfony/http-cache:dev-master@dev</summary>

```
	./composer.json has been updated
	Loading composer repositories with package information
	Updating dependencies (including require-dev)                           
	Dependency resolution completed in 0.046 seconds
	Your requirements could not be resolved to an installable set of packages.

	  Problem 1
	    - friendsofsymfony/http-cache dev-master requires symfony/event-dispatcher ^2.3||^3.0 -> satisfiable by symfony/event-dispatcher[2.3.x-dev, 2.4.x-dev, 2.5.x-dev, 2.6.x-dev, 2.7.x-dev, 2.8.x-dev, 3.0.x-dev, 3.1.x-dev, 3.2.x-dev, 3.3.x-dev, 3.4.x-dev, v2.3.0, v2.3.1, v2.3.10, v2.3.11, v2.3.12, v2.3.13, v2.3.14, v2.3.15, v2.3.16, v2.3.17, v2.3.18, v2.3.19, v2.3.2, v2.3.20, v2.3.21, v2.3.22, v2.3.23, v2.3.24, v2.3.25, v2.3.26, v2.3.27, v2.3.28, v2.3.29, v2.3.3, v2.3.30, v2.3.31, v2.3.32, v2.3.33, v2.3.34, v2.3.35, v2.3.36, v2.3.37, v2.3.38, v2.3.39, v2.3.4, v2.3.40, v2.3.41, v2.3.42, v2.3.5, v2.3.6, v2.3.7, v2.3.8, v2.3.9, v2.4.0, v2.4.0-BETA1, v2.4.0-BETA2, v2.4.0-RC1, v2.4.1, v2.4.10, v2.4.2, v2.4.3, v2.4.4, v2.4.5, v2.4.6, v2.4.7, v2.4.8, v2.4.9, v2.5.0, v2.5.0-BETA1, v2.5.0-BETA2, v2.5.0-RC1, v2.5.1, v2.5.10, v2.5.11, v2.5.12, v2.5.2, v2.5.3, v2.5.4, v2.5.5, v2.5.6, v2.5.7, v2.5.8, v2.5.9, v2.6.0, v2.6.0-BETA1, v2.6.0-BETA2, v2.6.1, v2.6.10, v2.6.11, v2.6.12, v2.6.13, v2.6.2, v2.6.3, v2.6.4, v2.6.5, v2.6.6, v2.6.7, v2.6.8, v2.6.9, v2.7.0, v2.7.0-BETA1, v2.7.0-BETA2, v2.7.1, v2.7.10, v2.7.11, v2.7.12, v2.7.13, v2.7.14, v2.7.15, v2.7.16, v2.7.17, v2.7.18, v2.7.19, v2.7.2, v2.7.20, v2.7.21, v2.7.22, v2.7.23, v2.7.24, v2.7.25, v2.7.26, v2.7.27, v2.7.28, v2.7.29, v2.7.3, v2.7.30, v2.7.31, v2.7.32, v2.7.33, v2.7.34, v2.7.35, v2.7.36, v2.7.37, v2.7.38, v2.7.4, v2.7.5, v2.7.6, v2.7.7, v2.7.8, v2.7.9, v2.8.0, v2.8.0-BETA1, v2.8.1, v2.8.10, v2.8.11, v2.8.12, v2.8.13, v2.8.14, v2.8.15, v2.8.16, v2.8.17, v2.8.18, v2.8.19, v2.8.2, v2.8.20, v2.8.21, v2.8.22, v2.8.23, v2.8.24, v2.8.25, v2.8.26, v2.8.27, v2.8.28, v2.8.29, v2.8.3, v2.8.30, v2.8.31, v2.8.4, v2.8.5, v2.8.6, v2.8.7, v2.8.8, v2.8.9, v3.0.0, v3.0.0-BETA1, v3.0.1, v3.0.2, v3.0.3, v3.0.4, v3.0.5, v3.0.6, v3.0.7, v3.0.8, v3.0.9, v3.1.0, v3.1.0-BETA1, v3.1.0-RC1, v3.1.1, v3.1.10, v3.1.2, v3.1.3, v3.1.4, v3.1.5, v3.1.6, v3.1.7, v3.1.8, v3.1.9, v3.2.0, v3.2.0-BETA1, v3.2.0-RC1, v3.2.0-RC2, v3.2.1, v3.2.10, v3.2.11, v3.2.12, v3.2.13, v3.2.14, v3.2.2, v3.2.3, v3.2.4, v3.2.5, v3.2.6, v3.2.7, v3.2.8, v3.2.9, v3.3.0, v3.3.0-BETA1, v3.3.0-RC1, v3.3.1, v3.3.10, v3.3.11, v3.3.12, v3.3.13, v3.3.2, v3.3.3, v3.3.4, v3.3.5, v3.3.6, v3.3.7, v3.3.8, v3.3.9, v3.4.0, v3.4.0-BETA1, v3.4.0-BETA2, v3.4.0-BETA3, v3.4.0-BETA4, v3.4.0-RC1, v3.4.0-RC2] but these conflict with your requirements or minimum-stability.
	    - friendsofsymfony/http-cache dev-master requires symfony/event-dispatcher ^2.3||^3.0 -> satisfiable by symfony/event-dispatcher[2.3.x-dev, 2.4.x-dev, 2.5.x-dev, 2.6.x-dev, 2.7.x-dev, 2.8.x-dev, 3.0.x-dev, 3.1.x-dev, 3.2.x-dev, 3.3.x-dev, 3.4.x-dev, v2.3.0, v2.3.1, v2.3.10, v2.3.11, v2.3.12, v2.3.13, v2.3.14, v2.3.15, v2.3.16, v2.3.17, v2.3.18, v2.3.19, v2.3.2, v2.3.20, v2.3.21, v2.3.22, v2.3.23, v2.3.24, v2.3.25, v2.3.26, v2.3.27, v2.3.28, v2.3.29, v2.3.3, v2.3.30, v2.3.31, v2.3.32, v2.3.33, v2.3.34, v2.3.35, v2.3.36, v2.3.37, v2.3.38, v2.3.39, v2.3.4, v2.3.40, v2.3.41, v2.3.42, v2.3.5, v2.3.6, v2.3.7, v2.3.8, v2.3.9, v2.4.0, v2.4.0-BETA1, v2.4.0-BETA2, v2.4.0-RC1, v2.4.1, v2.4.10, v2.4.2, v2.4.3, v2.4.4, v2.4.5, v2.4.6, v2.4.7, v2.4.8, v2.4.9, v2.5.0, v2.5.0-BETA1, v2.5.0-BETA2, v2.5.0-RC1, v2.5.1, v2.5.10, v2.5.11, v2.5.12, v2.5.2, v2.5.3, v2.5.4, v2.5.5, v2.5.6, v2.5.7, v2.5.8, v2.5.9, v2.6.0, v2.6.0-BETA1, v2.6.0-BETA2, v2.6.1, v2.6.10, v2.6.11, v2.6.12, v2.6.13, v2.6.2, v2.6.3, v2.6.4, v2.6.5, v2.6.6, v2.6.7, v2.6.8, v2.6.9, v2.7.0, v2.7.0-BETA1, v2.7.0-BETA2, v2.7.1, v2.7.10, v2.7.11, v2.7.12, v2.7.13, v2.7.14, v2.7.15, v2.7.16, v2.7.17, v2.7.18, v2.7.19, v2.7.2, v2.7.20, v2.7.21, v2.7.22, v2.7.23, v2.7.24, v2.7.25, v2.7.26, v2.7.27, v2.7.28, v2.7.29, v2.7.3, v2.7.30, v2.7.31, v2.7.32, v2.7.33, v2.7.34, v2.7.35, v2.7.36, v2.7.37, v2.7.38, v2.7.4, v2.7.5, v2.7.6, v2.7.7, v2.7.8, v2.7.9, v2.8.0, v2.8.0-BETA1, v2.8.1, v2.8.10, v2.8.11, v2.8.12, v2.8.13, v2.8.14, v2.8.15, v2.8.16, v2.8.17, v2.8.18, v2.8.19, v2.8.2, v2.8.20, v2.8.21, v2.8.22, v2.8.23, v2.8.24, v2.8.25, v2.8.26, v2.8.27, v2.8.28, v2.8.29, v2.8.3, v2.8.30, v2.8.31, v2.8.4, v2.8.5, v2.8.6, v2.8.7, v2.8.8, v2.8.9, v3.0.0, v3.0.0-BETA1, v3.0.1, v3.0.2, v3.0.3, v3.0.4, v3.0.5, v3.0.6, v3.0.7, v3.0.8, v3.0.9, v3.1.0, v3.1.0-BETA1, v3.1.0-RC1, v3.1.1, v3.1.10, v3.1.2, v3.1.3, v3.1.4, v3.1.5, v3.1.6, v3.1.7, v3.1.8, v3.1.9, v3.2.0, v3.2.0-BETA1, v3.2.0-RC1, v3.2.0-RC2, v3.2.1, v3.2.10, v3.2.11, v3.2.12, v3.2.13, v3.2.14, v3.2.2, v3.2.3, v3.2.4, v3.2.5, v3.2.6, v3.2.7, v3.2.8, v3.2.9, v3.3.0, v3.3.0-BETA1, v3.3.0-RC1, v3.3.1, v3.3.10, v3.3.11, v3.3.12, v3.3.13, v3.3.2, v3.3.3, v3.3.4, v3.3.5, v3.3.6, v3.3.7, v3.3.8, v3.3.9, v3.4.0, v3.4.0-BETA1, v3.4.0-BETA2, v3.4.0-BETA3, v3.4.0-BETA4, v3.4.0-RC1, v3.4.0-RC2] but these conflict with your requirements or minimum-stability.
	    - Installation request for friendsofsymfony/http-cache dev-master@dev -> satisfiable by friendsofsymfony/http-cache[dev-master].
```
</details>

**Comment by dbu** [comment]
this branch contains the changes JonEastman did, i just added more commits to it to see the actual build. 

if you want to test this branch, your composer command should say `composer -v require friendsofsymfony/http-cache:dev-symfony-4-updated@dev`. you are currently trying to install the master branch which does not allow symfony 4.

**Comment by pgnd** [comment]
@dbu 

> if you want to test this branch, your composer command should say composer -v require friendsofsymfony/http-cache:dev-symfony-4-updated@dev. you are currently trying to install the master branch which does not allow symfony 4.

Unless I'm installing from a local git instance of "symfony-4-updated", which I am ... where's version == "dev-master" (the default) is supposed to ref the current checked out branch's head.  i.e., in my case, == symfony-4-updated

(unless I'm mistaken, which is always an option)

as ref'd in the precursor to this bug:
```
  https://github.com/FriendsOfSymfony/FOSHttpCache/pull/389#issuecomment-347940576 
```

I've
```
cd /home/pgnd/src/FriendsOfSymfony-HttpCache
git branch -a | grep \*
	* symfony-4-updated

cd /home/pgnd/web/test
composer config repo.packagist false
composer config repositories.FriendsOfSymfony git /home/pgnd/src/FriendsOfSymfony-HttpCache

cat composer.json
	...
	"repositories": {
	    "FriendsOfSymfony": {
	        "type": "git",
	        "url": "/home/pgnd/src/FriendsOfSymfony-HttpCache"
	    },
	    "packagist": false
	}
	...
```
So, an install (with packagist DISABLED for clarity ...)
<details>
<summary>composer -vvv require friendsofsymfony/http-cache:dev-master@dev</summary>

```
Reading ./composer.json
Loading config file /home/pgnd/.composer/auth.json
Loading config file ./composer.json
Executing command (/home/pgnd/web/test): git branch --no-color --no-abbrev -v
Reading /home/pgnd/.composer/composer.json
Loading config file /home/pgnd/.composer/auth.json
Loading config file /home/pgnd/.composer/composer.json
Loading config file /home/pgnd/.composer/auth.json
Reading /home/pgnd/.composer/auth.json
Reading /home/pgnd/web/test/vendor/composer/installed.json
Reading /home/pgnd/.composer/vendor/composer/installed.json
Loading plugin Symfony\Flex\Flex
Running 1.5.2 (2017-09-11 16:59:25) with PHP 7.2.1-dev on Linux / 4.14.2-4.gb5596a5-default
Do not run Composer as root/super user! See https://getcomposer.org/home/pgnd for details
Reading /home/pgnd/.composer/cache/repo/https---symfony.sh/aliases.json from cache
Downloading https://symfony.sh/aliases.json
Writing /home/pgnd/.composer/cache/repo/https---symfony.sh/aliases.json into cache
Reading /home/pgnd/.composer/cache/repo/https---symfony.sh/versions.json from cache
Downloading https://symfony.sh/versions.json
./composer.json has been updated
Reading ./composer.json
Loading config file /home/pgnd/.composer/auth.json
Loading config file ./composer.json
Executing command (/home/pgnd/web/test): git branch --no-color --no-abbrev -v
Reading /home/pgnd/.composer/composer.json
Loading config file /home/pgnd/.composer/auth.json
Loading config file /home/pgnd/.composer/composer.json
Loading config file /home/pgnd/.composer/auth.json
Reading /home/pgnd/.composer/auth.json
Reading /home/pgnd/web/test/vendor/composer/installed.json
Reading /home/pgnd/.composer/vendor/composer/installed.json
Loading plugin Symfony\Flex\Flex_composer_tmp0
Loading composer repositories with package information
Executing command (/home/pgnd/src/FriendsOfSymfony-HttpCache): git show-ref --tags --dereference
Executing command (/home/pgnd/src/FriendsOfSymfony-HttpCache): git branch --no-color --no-abbrev -v
Executing command (/home/pgnd/src/FriendsOfSymfony-HttpCache): git branch --no-color
Executing command (/home/pgnd/src/FriendsOfSymfony-HttpCache): git show 'symfony-4-updated':'composer.json'
Executing command (/home/pgnd/src/FriendsOfSymfony-HttpCache): git log -1 --format=%at 'symfony-4-updated'
Reading composer.json of friendsofsymfony/http-cache (1.0.0)
Reading /home/pgnd/.composer/cache/repo/-usr-local-src-FriendsOfSymfony-HttpCache/8a842604322e728b629193739d388d0666a3c3f4 from cache
Importing tag 1.0.0 (1.0.0.0)
Reading composer.json of friendsofsymfony/http-cache (1.0.0-RC1)
Reading /home/pgnd/.composer/cache/repo/-usr-local-src-FriendsOfSymfony-HttpCache/060ca1769e04cc762a4ad4675c754f6fb3b47288 from cache
Importing tag 1.0.0-RC1 (1.0.0.0-RC1)
Reading composer.json of friendsofsymfony/http-cache (1.0.0-alpha1)
Reading /home/pgnd/.composer/cache/repo/-usr-local-src-FriendsOfSymfony-HttpCache/68dd4c2e5b0a5a862dafd1b66a0a856ae8367ce5 from cache
Importing tag 1.0.0-alpha1 (1.0.0.0-alpha1)
Reading composer.json of friendsofsymfony/http-cache (1.0.0-alpha2)
Reading /home/pgnd/.composer/cache/repo/-usr-local-src-FriendsOfSymfony-HttpCache/e0a547397c7ebc5fc6d6749002b04911aa60ac9f from cache
Importing tag 1.0.0-alpha2 (1.0.0.0-alpha2)
Reading composer.json of friendsofsymfony/http-cache (1.0.0-alpha3)
Reading /home/pgnd/.composer/cache/repo/-usr-local-src-FriendsOfSymfony-HttpCache/2d1ac0006ba34cc3b702202644e703777ea955f2 from cache
Importing tag 1.0.0-alpha3 (1.0.0.0-alpha3)
Reading composer.json of friendsofsymfony/http-cache (1.0.0-alpha4)
Reading /home/pgnd/.composer/cache/repo/-usr-local-src-FriendsOfSymfony-HttpCache/9f16da8bd204af553d0274a4e2a7ef8a833b57b8 from cache
Importing tag 1.0.0-alpha4 (1.0.0.0-alpha4)
Reading composer.json of friendsofsymfony/http-cache (1.0.0-beta1)
Reading /home/pgnd/.composer/cache/repo/-usr-local-src-FriendsOfSymfony-HttpCache/c63e30e74ac71a2854a7f5bcd8ffe940fed427ac from cache
Importing tag 1.0.0-beta1 (1.0.0.0-beta1)
Reading composer.json of friendsofsymfony/http-cache (1.0.1)
Reading /home/pgnd/.composer/cache/repo/-usr-local-src-FriendsOfSymfony-HttpCache/9c65814e72007d188673e7f413dfb0979fa5ff6f from cache
Importing tag 1.0.1 (1.0.1.0)
Reading composer.json of friendsofsymfony/http-cache (1.1.0)
Reading /home/pgnd/.composer/cache/repo/-usr-local-src-FriendsOfSymfony-HttpCache/9ca60d9beed3e3d3849c736788176e78565a2154 from cache
Importing tag 1.1.0 (1.1.0.0)
Reading composer.json of friendsofsymfony/http-cache (1.1.1)
Reading /home/pgnd/.composer/cache/repo/-usr-local-src-FriendsOfSymfony-HttpCache/480615b3c80472d62646d66f3e3495a3e6a9ab87 from cache
Importing tag 1.1.1 (1.1.1.0)
Reading composer.json of friendsofsymfony/http-cache (1.1.2)
Reading /home/pgnd/.composer/cache/repo/-usr-local-src-FriendsOfSymfony-HttpCache/4e227b47754e2505fc6cf21cd2c4e98d625a3e11 from cache
Importing tag 1.1.2 (1.1.2.0)
Reading composer.json of friendsofsymfony/http-cache (1.2.0)
Reading /home/pgnd/.composer/cache/repo/-usr-local-src-FriendsOfSymfony-HttpCache/8da5db9a0b5b3af426947eb501bf45c82cc844ca from cache
Importing tag 1.2.0 (1.2.0.0)
Reading composer.json of friendsofsymfony/http-cache (1.3.0)
Reading /home/pgnd/.composer/cache/repo/-usr-local-src-FriendsOfSymfony-HttpCache/d5e382f59504cf373b5493f779f604b5fb568de1 from cache
Importing tag 1.3.0 (1.3.0.0)
Reading composer.json of friendsofsymfony/http-cache (1.3.1)
Reading /home/pgnd/.composer/cache/repo/-usr-local-src-FriendsOfSymfony-HttpCache/fa6cd4a87715d15eb283f02fdea6ed6b2dda6276 from cache
Importing tag 1.3.1 (1.3.1.0)
Reading composer.json of friendsofsymfony/http-cache (1.3.2)
Reading /home/pgnd/.composer/cache/repo/-usr-local-src-FriendsOfSymfony-HttpCache/6269866289ce9b80b7e569b4cffdc829fbe7452b from cache
Importing tag 1.3.2 (1.3.2.0)
Reading composer.json of friendsofsymfony/http-cache (1.4.0)
Reading /home/pgnd/.composer/cache/repo/-usr-local-src-FriendsOfSymfony-HttpCache/54117e25b9a0ac4568d33f349d62ef83d2301b76 from cache
Importing tag 1.4.0 (1.4.0.0)
Reading composer.json of friendsofsymfony/http-cache (1.4.1)
Reading /home/pgnd/.composer/cache/repo/-usr-local-src-FriendsOfSymfony-HttpCache/eb24688cfc4a553061add8196caf1783c45d450a from cache
Importing tag 1.4.1 (1.4.1.0)
Reading composer.json of friendsofsymfony/http-cache (1.4.2)
Reading /home/pgnd/.composer/cache/repo/-usr-local-src-FriendsOfSymfony-HttpCache/783f3918d85263e444db02a9e42d0f42aa7c1f43 from cache
Importing tag 1.4.2 (1.4.2.0)
Reading composer.json of friendsofsymfony/http-cache (1.4.3)
Reading /home/pgnd/.composer/cache/repo/-usr-local-src-FriendsOfSymfony-HttpCache/de883f0a76d78935278942e25e093418bfbe6c26 from cache
Importing tag 1.4.3 (1.4.3.0)
Reading composer.json of friendsofsymfony/http-cache (1.4.4)
Reading /home/pgnd/.composer/cache/repo/-usr-local-src-FriendsOfSymfony-HttpCache/e528b0fb046d3f329b10951201a5b0cab60e96e1 from cache
Importing tag 1.4.4 (1.4.4.0)
Reading composer.json of friendsofsymfony/http-cache (2.0.0)
Reading /home/pgnd/.composer/cache/repo/-usr-local-src-FriendsOfSymfony-HttpCache/bd0ce2938f86c9e1db339f5ad2b0bf9c593fe34c from cache
Importing tag 2.0.0 (2.0.0.0)
Reading composer.json of friendsofsymfony/http-cache (2.0.0-alpha1)
Reading /home/pgnd/.composer/cache/repo/-usr-local-src-FriendsOfSymfony-HttpCache/9a4519ca202c324fcc3214d1cc1689397558898c from cache
Importing tag 2.0.0-alpha1 (2.0.0.0-alpha1)
Reading composer.json of friendsofsymfony/http-cache (2.0.0-beta1)
Reading /home/pgnd/.composer/cache/repo/-usr-local-src-FriendsOfSymfony-HttpCache/ac4303bc7c5c996da932a744223084f020f0d99b from cache
Importing tag 2.0.0-beta1 (2.0.0.0-beta1)
Reading composer.json of friendsofsymfony/http-cache (2.0.0-beta2)
Reading /home/pgnd/.composer/cache/repo/-usr-local-src-FriendsOfSymfony-HttpCache/42ff6dd772bc4fb24c1d56332d8237715cbd2d1b from cache
Importing tag 2.0.0-beta2 (2.0.0.0-beta2)
Reading composer.json of friendsofsymfony/http-cache (2.0.0-beta3)
Importing tag 2.0.0-beta3 (2.0.0.0-beta3)
Reading composer.json of friendsofsymfony/http-cache (2.0.1)
Reading /home/pgnd/.composer/cache/repo/-usr-local-src-FriendsOfSymfony-HttpCache/ecc458282b8a6f75387c3cb4705615b0f5b6d6a8 from cache
Importing tag 2.0.1 (2.0.1.0)
Reading composer.json of friendsofsymfony/http-cache (2.0.2)
Reading /home/pgnd/.composer/cache/repo/-usr-local-src-FriendsOfSymfony-HttpCache/63837b29670094539fdda35cb887cbfc57798c6f from cache
Importing tag 2.0.2 (2.0.2.0)
Reading composer.json of friendsofsymfony/http-cache (master)
Reading /home/pgnd/.composer/cache/repo/-usr-local-src-FriendsOfSymfony-HttpCache/5ddf38c348f421907ba711c3fa4a77c25a536802 from cache
Importing branch master (dev-master)
Reading composer.json of friendsofsymfony/http-cache (symfony-4-updated)
Reading /home/pgnd/.composer/cache/repo/-usr-local-src-FriendsOfSymfony-HttpCache/b53772792ee0b88f7e28b253b9813566b4797f93 from cache
Importing branch symfony-4-updated (dev-symfony-4-updated)
Updating dependencies (including require-dev)
Resolving dependencies through SAT
Dependency resolution completed in 0.001 seconds
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Installation request for friendsofsymfony/http-cache dev-master@dev -> satisfiable by friendsofsymfony/http-cache[dev-master].
    - friendsofsymfony/http-cache dev-master requires symfony/event-dispatcher ^2.3||^3.0 -> no matching package found.

Potential causes:
 - A typo in the package name
 - The package is not available in a stable-enough version according to your minimum-stability setting
   see <https://getcomposer.org/doc/04-schema.md#minimum-stability> for more details.

Read <https://getcomposer.org/doc/articles/troubleshooting.md> for further common problems.

Installation failed, reverting ./composer.json to its original content.
```
</details>

**Comment by pgnd** [comment]
direct checkout from this repo on github installs OK.

local checkout of, supposedly, an identical repo/branch, does not (fails as above).

OTHER projects (non-SF4-involved ...) I'm able to install using a local instance.

So _something_ is different about this local instance; I suspect another FOS dependency that's not properly resolved by my local setup.  Will figure that out separately.

For now, for this issue ... switching to direct checkout.

**Comment by pgnd** [comment]
> no x-cache response header added

With
```
	SF4
	HttpCache:dev-symfony-4-updated
	php 7.2.1
	varnish 5.2.1
```
and some migration fussing, I'm able to get an initial response, currently only a MISS, but I *do* see x-cache *header*

<details>
<summary>Request Header</summary>

```
	Host: dev.loc
	User-Agent: Mozilla/5.0 (X11; Linux x86_64; rv:58.0) Gecko/20100101 Firefox/58.0
	Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
	Accept-Language: en-US,en;q=0.5
	Accept-Encoding: gzip, deflate, br
	DNT: 1
	Connection: keep-alive
	Upgrade-Insecure-Requests: 1
	Pragma: no-cache
	Cache-Control: no-cache
```
</details>

<details>
<summary>Response Header</summary>

```
	date: Sat, 02 Dec 2017 00:01:24 GMT
	content-type: text/html; charset=UTF-8
	vary: Accept-Encoding, Accept-Encoding
	cache-control: no-cache, private
	x-debug-token: dc5ab0
	x-debug-token-link: https://dev.loc/_profiler/dc5ab0
	x-ua-compatible: IE=edge,chrome=1
	x-xss-protection: 1; mode=block
	x-frame-options: sameorigin
	age: 0
	x-cache: MISS
	x-cache-hits: 0
	secure: Server
	strict-transport-security: max-age=315360000; includeSubDomains; preload
	public-key-pins: pin-sha256="62uCadBQ5OtAJKQ7z/X3LWIy3bPOgMJobMV+kkuFcJY="; pin-sha256="iWTndXWV48mg4cH1LxO49C69h2CQpmYfUYtWPHXdo8g="; max-age=10; includeSubdomains;
	x-content-type-options: nosniff, nosniff
	content-encoding: br
	X-Firefox-Spdy: h2
```
</details>
<br>  &
<details>
<summary>varnishtop</summary>

```
	...
   130.60 ReqHeader      Cookie:
   130.60 ReqUnset       Cookie:
    23.88 ReqHeader      Host: dev.loc
    17.95 ReqHeader      X-Forwarded-Port: 443
    17.95 ReqHeader      X-Forwarded-For: 10.0.0.1
    12.02 ReqMethod      GET
    12.02 VCL_call       HASH
    12.02 VCL_call       RECV
    12.02 ReqHeader      dnt: 1
    12.02 ReqProtocol    HTTP/1.0
    12.02 VCL_return     lookup
    12.02 ReqHeader      HTTPS: on
    12.02 VCL_return     deliver
    12.02 ReqHeader      X-Purge: Yes
    12.02 ReqHeader      X-Real-IP: 10.0.0.1
    12.02 ReqHeader      pragma: no-cache
    12.02 ReqHeader      Connection: close
    12.02 ReqHeader      Accept-Encoding: gzip
    12.02 ReqHeader      X-Client-Verify: SUCCESS
    12.02 ReqHeader      cache-control: no-cache
    12.02 ReqHeader      X-Forwarded-Proto: https
    12.02 ReqHeader      accept-language: en-US,en;q=0.5
    12.02 ReqUnset       Host: dev.loc
    12.02 ReqHeader      user-agent: Mozilla/5.0 (X11; Linux x86_64; rv:58.0) Gecko/20100101
    11.00 ReqHeader      referer: https://dev.loc/
     8.71 ReqHeader      accept: */*
     7.29 VCL_return     fetch
     6.10 VCL_call       PURGE
     6.10 VCL_return     hash
     6.10 RespProtocol   HTTP/1.1
     6.10 VCL_call       DELIVER
     6.10 VCL_return     purge
     6.10 VCL_return     restart
     6.10 Begin          sess 0 HTTP/1
     6.10 RespHeader     Connection: close
     6.10 ReqUnset       X-Forwarded-Port: 443
     6.10 ReqUnset       X-Forwarded-For: 10.0.0.1
     6.10 RespHeader     Via: 1.1 varnish (Varnish/5.2)
     6.10 RespUnset      Via: 1.1 varnish (Varnish/5.2)
```
</details>
<br>Now to get some cacheable content and poke around ...

**Comment by Toflar** [comment]
> now we see the real problems with symfony 4 compatibility:
> 
>     no x-cache response header added
>     usage of Request::isMethodSafe in a way that triggers an exception

Second one caused also the first ones because if an exception is thrown, no x-cache headers are added ;-) See https://github.com/FriendsOfSymfony/FOSHttpCache/pull/395

File: src/SymfonyCache/RefreshListener.php [hunk 1]
```diff
@@ -44,7 +44,7 @@ public static function getSubscribedEvents()
     public function handleRefresh(CacheEvent $event)
     {
         $request = $event->getRequest();
-        if (!$request->isMethodSafe()
+        if (!$request->isMethodSafe(false)
```
Comment (dbu) [inline src/SymfonyCache/RefreshListener.php]: looking at https://github.com/symfony/symfony/commit/34e7b956dd46cc03973ba90ce93523557a26a3e0 i think we misused this method to see if the request method is cacheable, so we should instead check if the method is cacheable. will have a look.

Comment (Toflar) [reply src/SymfonyCache/RefreshListener.php]: Yeah, indeed I was unsure myself. Should've mentioned that, sorry!

File: src/SymfonyCache/UserContextListener.php [hunk 2]
```diff
@@ -104,9 +104,9 @@ public function preHandle(CacheEvent $event)
                 return;
             }
 
-            // We must not call $request->isMethodSafe() here, because this will mess up with later configuration of the `http_method_override` that is set onto the request by the Symfony FrameworkBundle.
+            // We must not call $request->isMethodCacheable() here, because this will mess up with later configuration of the `http_method_override` that is set onto the request by the Symfony FrameworkBundle.
             // See http://symfony.com/doc/current/reference/configuration/framework.html#configuration-framework-http-method-override
-            if (in_array($request->getRealMethod(), ['GET', 'HEAD', 'OPTIONS', 'TRACE']) && $hash = $this->getUserHash($event->getKernel(), $request)) {
+            if (in_array($request->getRealMethod(), ['GET', 'HEAD']) && $hash = $this->getUserHash($event->getKernel(), $request)) {
```
Comment (dbu) [inline src/SymfonyCache/UserContextListener.php]: i think this is what we meant to do. @Toflar do you agree or did i somehow misinterpret what we do? we want to know if the request method is cacheable. OPTIONS and TRACE usually are not cacheable.

Comment (Toflar) [reply src/SymfonyCache/UserContextListener.php]: I agree, `OPTIONS` and `TRACE` must not be cached acccording to spec. Perfect 👍

**Comment by dbu** [comment]
thanks a lot for the help everybody!

**Pull request merged**